lib/update/cluster/engine.go

```diff
@@ -0,0 +1,325 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"github.com/gravitational/gravity/lib/app"
+	"github.com/gravitational/gravity/lib/checks"
+	"github.com/gravitational/gravity/lib/defaults"
+	"github.com/gravitational/gravity/lib/fsm"
+	"github.com/gravitational/gravity/lib/ops"
+	"github.com/gravitational/gravity/lib/pack"
+	"github.com/gravitational/gravity/lib/storage"
+	"github.com/gravitational/gravity/lib/update"
+	"github.com/gravitational/gravity/lib/users"
+	"github.com/gravitational/gravity/lib/utils"
+
+	"github.com/gravitational/trace"
+	"github.com/pborman/uuid"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+)
+
+// New returns new updater for the specified configuration
+func New(ctx context.Context, config Config) (*update.Updater, error) {
+	machine, err := newMachine(ctx, config)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	updater, err := update.NewUpdater(ctx, config.Config, machine)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return updater, nil
+}
+
+// checkAndSetDefaults validates FSM config and sets defaults
+func (c *Config) checkAndSetDefaults() error {
+	if err := c.Config.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	if c.Spec == nil {
+		c.Spec = fsmSpec(*c)
+	}
+	return nil
+}
+
+// Config is the FSM configuration
+type Config struct {
+	update.Config
+	// HostLocalBackend is the host-local backend that stores bootstrap configuration
+	// like DNS, logins etc.
+	HostLocalBackend storage.Backend
+	// Packages is the local package service
+	Packages pack.PackageService
+	// ClusterPackages is the package service that talks to cluster API
+	ClusterPackages pack.PackageService
+	// HostLocalPackages is the host-local package service that contains package
+	// metadata used for updates
+	HostLocalPackages pack.PackageService
+	// Apps is the cluster apps service
+	Apps app.Applications
+	// Client is the cluster Kubernetes client
+	Client *kubernetes.Clientset
+	// Users is the cluster identity service
+	Users users.Identity
+	// Spec is used to retrieve a phase executor, allows
+	// plugging different phase executors during tests
+	Spec fsm.FSMSpecFunc
+}
+
+// newMachine returns a new FSM instance
+func newMachine(ctx context.Context, c Config) (*fsm.FSM, error) {
+	err := c.checkAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	logger := logrus.WithFields(logrus.Fields{
+		trace.Component: "fsm:update",
+	})
+	engine, err := newEngine(ctx, c, logger)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	fsm, err := fsm.New(fsm.Config{
+		Engine: engine,
+		Logger: logger,
+		Runner: c.Runner,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return fsm, nil
+}
+
+// engine is the update FSM engine
+type engine struct {
```

Comment:
This can be made to use the same rollingupdate package as used bu `RuntimeEnvironment` / `ClusterConfiguration` operations - this is something for a later PR though.